### PR TITLE
feat(compute,api,update-server): Move system configs out of Dockerfile

### PR DESCRIPTION
**Pull request by sfoster1**
## overview

Overhaul of the container configuration and initialization to remove as much as
possible from the Dockerfile and the volatile storage, preferring instead to
live in files that are included in the api server wheel and unpacked to /data
during initialization.

Most configuration and setup scripts are no longer COPYd into /usr/local/bin.
Instead, they are packed into the api server’s wheel (opentrons) in a
/resources subdirectory. This resources subdirectory is recursive-included
into the manifest, so putting a file into the directory will include it in the
wheel.

Because wheels do not run arbitrary code when they are installed, there is some
external tooling to get the files out of wheel/resources and into their eventual
home in /data/system.

* Provisioning

The compute/find_module_path.py is an executable python script that uses
importlib to find the location of the opentrons module as Python would do (so it
always gets the right one) without importing opentrons, which has significant
side effects.

The opentrons/api/opentrons/resources/scripts/provision-api-resources script
uses find_module_path to find the current opentrons module and copy everything
in /resources in the package to /data/system (the OT_CONFIG_DIR). This includes
all the system configurations and scripts that were previously in compute/.

The provision script is designed to be invoked during updates. This happens
during the first boot of the container (more on this later) and during a
software update via the update server.

In the case of the update-server driven update, because we assume a user is
never updating to 3.3 from 3.0 on a 3.0 Dockerfile, we can assume the provision
script is in /data/system/scripts and is therefore in the path. The update
server therefore invokes the provision script after installing a new api server,
the provision script finds the right (newly-installed) opentrons module, and
updates /data/resources with the new scripts.

In the case of the first container boot, we rely on the one piece of
initialization left in compute/. compute/container_setup.sh does the first-boot
check that setup.sh used to do, and in addition to removing old api server
installations and updating the cached container ID it runs the provision script
- this time from the module installed in /usr/local/lib.

* Container Initialization

The Dockerfile CMD has been slightly simplified by symlinking the environment
file into /etc/profile.d. This requires the use of bash’s -l flag to get a login
shell in the docker container.

Because the symlinks to the environment must be present when the shell starts,
the environment script (opentrons/api/opentrons/resources/ot-environ.sh) is
linked into /etc/profile.d by the Dockerfile (the rest of the configuration file
symlinks are in opentrons/api/opentrons/resources/scripts/setup.sh). It is also
linked twice - once from /data/system, which will be empty when the container
boots for the first time; and once from the system installation of the api
server in /usr/local/lib, as a fallback. The one-time environment variables
ensure it is only executed once.

Then, the container setup script runs. In most cases it doesn’t do much, but see
above for the behavior during the first container boot.

Since /data/system/scripts is in the path, we still call setup.sh and start.sh.
In addition to what it used to do, setup.sh now makes symlinks for most of the
configuration files we had been COPYing into Docker. These are now in
/data/system. start.sh is pretty similar to before.

* Container Building

The Dockerfile is now parameterized to make building a local container slightly
easier. Invoking docker build with no arguments builds a container for the
pi (it has to, since this is what Resin does). Invoking docker build with the
arguments to change the base image and clear RUNNING_ON_PI will build a
container for local machines. To make this easier, there is a new top-level
Makefile target, api-local-container, that invokes docker build with the correct
arguments. Note that the local container still doesn’t 100% work; it needs more
love to get over things like not having a dbus socket on all hosts.

## changelog

- Move container configuration and initialization to opentrons/api/opentrons/resources as much as possible
- Copy the data in /resources to /data/system on first container boot and during update (in the update server)
- Sweet new motd
- Removed some of the nmcli commands used for janitoring the static ipv6 routes
- Removed the nginx root server block and deleted the static files it was
- serving
- hardcoded some ports to get away from infinite env vars we can’t trust
- Fixed an issue in the update server where it was sending the repr() of
  tracebacks in 500 messages for /server/update. Now it sends the result of traceback.format_tb()

Note that Dockerfiles without these changes will not get the new configuration and startup, even if an api server is properly installed and provisioned, since the environment script won't be sourced.

## review requests

Test:

- [x] A robot that is on 3.0.0 and has its resin-updates.lock file deleted successfully downloads this container from resin, boots, and has all its old config data from the usb driver
- [x] A robot that has this docker file, update server, and api server can successfully be updated (by run app) to a new version, including provisioning (usually check by modifying eg /data/system/ot-environ.sh and check that the modifications are reverted after update)
- [x] Unless specifically erased by updating, changes in /data/system should persist and take effect after a reboot. For instance, a change to /data/system/ot-environ.sh should be present even after a reboot. Note that the resin.io web shell doesn't actually use a login shell, so it may appear as if configurations are erased. To see the correct environment, do bash -l in the resin webshell (ssh and make term do this by default)
- Also that everything still works: 
  - [x] the api and update server respond
  - [x] Jupyter works
  - [x] it connects to wifi
  - [ ] it connects via usb 

@andySigler Would you mind double-checking ot-environ.sh and the contents of api/opentrons/resources and making sure that the stuff you need for the factory is OK?

Everybody else Let me know if there's more testing I should be doing.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -59,13 +42,11 @@ RUN pip install --force-reinstall \
 # Copy server files and data into the container. Note: any directories that
 # you wish to copy into the container must be excluded from the .dockerignore
 # file, or you will encounter a copy error
-ENV LABWARE_DEF /etc/labware
-ENV AUDIO_FILES /etc/audio
-ENV USER_DEFN_ROOT /data/user_storage/opentrons_data/labware
+
+COPY compute/container_setup.sh /usr/local/bin/container_setup.sh
```
Comment (mcous) [inline Dockerfile]: [Style nitpick] All the other paths in this file have a leading `./`

File: Dockerfile [hunk 2]
```diff
@@ -91,48 +72,15 @@ RUN ln -sf /dev/stdout /var/log/nginx/access.log && \
...
-RUN gpg --import opentrons.asc && rm opentrons.asc
-
-# Everything you want in /usr/local/bin goes into compute/scripts
-COPY ./compute/scripts/* /usr/local/bin/
-
-# All configuration files live in compute/etc and dispatched here
-COPY ./compute/conf/inetd.conf /etc/
-COPY ./compute/conf/nginx.conf /etc/nginx/nginx.conf
-COPY ./compute/static /usr/share/nginx/html
+# COPY ./compute/opentrons.asc .
```
Comment (mcous) [inline Dockerfile]: Is it ok to remove these lines entirely?

File: api/opentrons/__init__.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import os
 import sys
 import json
+import shutil
```
Comment (mcous) [inline api/opentrons/__init__.py]: Unused import

File: api/opentrons/resources/ot-environ.sh [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+# /etc/profile.d by the Dockerfile.
+
+if [ -z $OT_ENVIRON_SET_UP ]; then
+    echo "[ $0 ] Configuring environment"
+
+    arg_running_on_pi=$1
+    if [[ -z "$RUNNING_ON_PI" ]]
+    then
+        echo '[ $0 ] Forcing virtual smoothie board'
+        export ENABLE_VIRTUAL_SMOOTHIE=1
```
Comment (mcous) [inline api/opentrons/resources/ot-environ.sh]: Are we sure we want to do this? Running on a machine that isn't a Pi that is still connected to a real smoothie board seems like a valid dev use-case

File: api/opentrons/resources/ot-environ.sh [hunk 5]
```diff
@@ -0,0 +1,41 @@
+# Set up the environment for the OT2.
+# This is sourced by the system login shell profile by symlinks placed in
+# /etc/profile.d by the Dockerfile.
+
+if [ -z $OT_ENVIRON_SET_UP ]; then
+    echo "[ $0 ] Configuring environment"
+
+    arg_running_on_pi=$1
```
Comment (mcous) [inline api/opentrons/resources/ot-environ.sh]: This variable doesn't appear to be used

File: api/opentrons/resources/ot-environ.sh [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    # This is used by D-Bus clients such as Network Manager cli, announce_mdns
+    # connecting to Host OS services
+    export DBUS_SYSTEM_BUS_ADDRESS=unix:path=/host/run/dbus/system_bus_socket
+    export PYTHONPATH=$PYTHONPATH:/data/packages/usr/local/lib/python3.6/site-packages
+    export PATH=$PATH:/data/packages/usr/local/bin:$OT_CONFIG_PATH/scripts
+
+    # TODO(seth, 8/15/2018): These are almost certainly unused and should be hardcoded
+    # if they are in fact still used
+    export OT_SETTINGS_DIR=""
+    export OT_SERVER_PORT=31950
```
Comment (mcous) [inline api/opentrons/resources/ot-environ.sh]: This env var was only ever used by the Dockerfile to build the NGINX conf. If we're not going to use it maybe we should omit it?

File: api/opentrons/resources/ot-environ.sh [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    export OT_SERVER_UNIX_SOCKET_PATH=/tmp/aiohttp.sock
```
Comment (mcous) [inline api/opentrons/resources/ot-environ.sh]: This PR might not be the time, but should we consider either ditching the Unix socket or aligning everything to use Unix sockets? At the very least, this is a really unhelpful socket name that we've been using

File: api/opentrons/resources/scripts/setup.sh [hunk 8]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+
+echo "[ $0 ] API server setup beginning"
+
+if [ ! -z $RUNNING_ON_PI ] ; then
```
Comment (mcous) [inline api/opentrons/resources/scripts/setup.sh]: [Nitpick, no change requested] Any reason to use `! -z $RUNNING_ON_PI` here over `-n "$RUNNING_ON_PI"`?

File: api/opentrons/resources/scripts/start.sh [hunk 9]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
+echo "[ $0 ] API server starting"
+# mdns announcement
+if [ ! -z $RUNNING_ON_PI ]; then
+    echo "[ $0 ] MDNS beginning"
```
Comment (mcous) [inline api/opentrons/resources/scripts/start.sh]: Out of curiosity, why disable mDNS advertisements on non-Pis?

File: api/opentrons/resources/scripts/start.sh [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+config_path=`python -c "from opentrons import config; print(config.get_config_index().get('deckCalibrationFile'))"`
+
+if [ ! -e "$config_path" ]; then
+    echo $config_path
+    echo "[ $0 ] Config file not found. Please perform factory calibration and then restart robot"
+fi
+
+export ENABLE_NETWORKING_ENDPOINTS=true
+echo "[ $0 ] Starting Opentrons API server"
+python -m opentrons.server.main -U /tmp/aiohttp.sock opentrons.server.main:init
```
Comment (mcous) [inline api/opentrons/resources/scripts/start.sh]: We should probably use the `OT_SERVER_UNIX_SOCKET_PATH` environment variable here

File: compute/README.md [hunk 11]
```diff
@@ -44,6 +44,10 @@ robot. The script must:
 Scripts in that directory will be executed in alphabetical order. The most straight-forward way to guarantee exectution
 order is to prefix file names with numbers (e.g.: `00-configure-wifi` and `05-do-something-else`).
 
+### System Configuration and Initialization
```
Comment (mcous) [inline compute/README.md]: [Nice to have] While we're at improving documentation, does the top-level file `RESIN_README.md` mean anything anymore? If not, might be nice to ditch it in this PR

File: api/opentrons/resources/ot-environ.sh (hunk 4 above)
Comment (sfoster1) [reply api/opentrons/resources/ot-environ.sh]: Sure

File: api/opentrons/resources/ot-environ.sh (hunk 6 above)
Comment (sfoster1) [reply api/opentrons/resources/ot-environ.sh]: That makes sense. I had a chat with Ben and we decided to bias towards keeping everything for fear of breaking e.g. the factory; but if there are some variables you know aren't used please point them out.

File: api/opentrons/resources/scripts/setup.sh (hunk 8 above)
Comment (sfoster1) [reply api/opentrons/resources/scripts/setup.sh]: Slowly re-remembering bash conditionals, mostly.

File: api/opentrons/resources/scripts/start.sh (hunk 9 above)
Comment (sfoster1) [reply api/opentrons/resources/scripts/start.sh]: This is one of a category of things in the container that only work at the very least on linux hosts and quite possibly only on the pi. I haven't been consistent with the rest of them, I think what we should do is create an issue to actually get a container running well on (osx, windows) docker.

File: compute/README.md (hunk 11 above)
Comment (sfoster1) [reply compute/README.md]: Will do

**Pull request merged**